### PR TITLE
Improve Artist docstrings

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

- Adopting our current docstring conventions.
- Improved a number of descriptions.
- Reorder the methods in the overview `artist_api.rst`. There's nothing added or removed in this file.
  *Motivation:* Since the top-level grouping is semantically, I try to order related functions within the sections in a logical way. In particular I move getters and setters next to each other.

There's still a bit more to do within this module, but I think the PR is big enough for now.

### CODE REVIEW

File: lib/matplotlib/artist.py [hunk 1]
```diff
@@ -284,11 +286,28 @@ def get_tightbbox(self, renderer):
...
-        Adds a callback function that will be called whenever one of
-        the :class:`Artist`'s properties changes.
+        Add a callback function that will be called whenever one of the
+        `.Artist`'s properties changes.
 
-        Returns an *id* that is useful for removing the callback with
-        :meth:`remove_callback` later.
+        Parameters
+        ----------
+        func : Callable[Artist]
```
Comment (anntzer) [inline lib/matplotlib/artist.py]: Technically that's `Callable[[Artist], Any]` (I think).

File: lib/matplotlib/artist.py [hunk 2]
```diff
@@ -297,13 +316,11 @@ def add_callback(self, func):
 
     def remove_callback(self, oid):
         """
-        Remove a callback based on its *id*.
-
-        .. seealso::
-
-            :meth:`add_callback`
-               For adding callbacks
+        Remove a callback based on its observer *id*.
```
Comment (anntzer) [inline lib/matplotlib/artist.py]: `oid`, not `id`

File: lib/matplotlib/artist.py [hunk 3]
```diff
@@ -220,8 +223,8 @@ def axes(self, new_axes):
     @property
     def stale(self):
         """
-        If the artist is 'stale' and needs to be re-drawn for the output to
-        match the internal state of the artist.
+        A boolean flag indicating if the artist is 'stale' and needs to be
```
Comment (anntzer) [inline lib/matplotlib/artist.py]: Actually I prefer the old version (after s/If/Whether).

File: lib/matplotlib/artist.py [hunk 4]
```diff
@@ -370,18 +391,28 @@ def hitlist(self, event):
         return L
 
     def get_children(self):
-        """
-        Return a list of the child :class:`Artist`s this
-        :class:`Artist` contains.
-        """
+        """Return a list of the child `.Artist`\ s this `.Artist` contains."""
```
Comment (anntzer) [inline lib/matplotlib/artist.py]: I think this should be a raw string with no space after the backslash.

File: lib/matplotlib/artist.py [hunk 5]
```diff
@@ -390,40 +421,61 @@ def contains(self, mouseevent):
...
-        If the mouse event is over the artist, return *hit* = *True*
-        and *props* is a dictionary of properties you want returned
-        with the contains test.
+        The provided callable replaces the default `.contains` method
+        of the artist.
 
         Parameters
         ----------
-        picker : callable
+        picker : Callable[Artist, MouseEvent]
```
Comment (anntzer) [inline lib/matplotlib/artist.py]: See above.  I would also not bother repeating the signature just below.  (Or we can just decide to write type annotations in "signature" form as below; either way I don't think repeating the same info twice in slightly different formats really helps here.)

File: lib/matplotlib/artist.py [hunk 6]
```diff
@@ -566,19 +627,19 @@ def get_sketch_params(self):
 
         Returns
         -------
-        sketch_params : tuple or `None`
+        sketch_params : tuple or None
 
             A 3-tuple with the following elements:
 
-              * `scale`: The amplitude of the wiggle perpendicular to the
+              - *scale*: The amplitude of the wiggle perpendicular to the
```
Comment (anntzer) [inline lib/matplotlib/artist.py]: Unindent the list.  Probably also remove the blank lines in between.

File: lib/matplotlib/artist.py [hunk 7]
```diff
@@ -1031,20 +1090,26 @@ def findobj(self, match=None, include_self=True):
         """
         Find artist objects.
 
-        Recursively find all :class:`~matplotlib.artist.Artist` instances
-        contained in self.
-
-        *match* can be
+        Recursively find all `.Artist` instances contained in self.
```
Comment (anntzer) [inline lib/matplotlib/artist.py]: `in *self*` perhaps

File: lib/matplotlib/artist.py [hunk 8]
```diff
@@ -1031,20 +1090,26 @@ def findobj(self, match=None, include_self=True):
...
+            A filter criterion for the matches. This can be
 
-          - function with signature ``boolean = match(artist)``
-            used to filter matches
+            - *None*: Return all objects contained in artist.
+            - A function with signature ``def match(artist: Artist) -> bool``.
+              The result will only contain artists for which the function
+              returns *True*.
+            - A class instance: e.g., `.Line2D`. The result will only contain
+              artists of this class or it's subclasses (``isinstance`` check).
```
Comment (anntzer) [inline lib/matplotlib/artist.py]: its

File: lib/matplotlib/artist.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/artist.py]: Reverted to just `callback`. In such cases, I will write the required signature as a type-annotated example in such cases.

IMO `def func(artist: Artist) -> Any` is easier to understand than `Callable[[Artist], Any]`.

File: lib/matplotlib/artist.py (hunk 2 above)
Comment (timhoffm) [reply lib/matplotlib/artist.py]: Changed to "observer id" without markup. "observer *oid*" would be sort of double.

File: lib/matplotlib/artist.py (hunk 5 above)
Comment (timhoffm) [reply lib/matplotlib/artist.py]: Switching to just "callable" and giving an example signature below (see also comment above).

File: lib/matplotlib/artist.py [hunk 9]
```diff
@@ -312,16 +329,23 @@ def remove_callback(self, oid):
 
     def pchanged(self):
         """
-        Fire an event when property changed, calling all of the
-        registered callbacks.
+        Call all of the registered callbacks.
+
+        This function is triggered internally when a property changed.
```
Comment (dstansby) [inline lib/matplotlib/artist.py]: "property changed" > "property is changed"

**Pull request merged**